odxtools/encodestate.py

```diff
@@ -28,31 +37,36 @@ class EncodeState:
     length_keys: Dict[str, int] = field(default_factory=dict)
 
     #: Mapping from the short name of a table-key parameter to the
-    #: corresponding row of the table (specified by TableKeyParameter)
-    table_keys: Dict[str, "TableRow"] = field(default_factory=dict)
+    #: short name of the corresponding row of the table (specified by
+    #: TableKeyParameter)
+    table_keys: Dict[str, str] = field(default_factory=dict)
+
+    #: The cursor position where a given length- or table key is located
+    #: in the PDU
+    key_pos: Dict[str, int] = field(default_factory=dict)
 
     #: Flag whether we are currently the last parameter of the PDU
-    #: (needed for MinMaxLengthType)
-    is_end_of_pdu: bool = False
+    #: (needed for MinMaxLengthType, EndOfPduField, etc.)
+    is_end_of_pdu: bool = True
 
-    def emplace_atomic_value(self,
-                             new_data: bytes,
-                             param_name: str,
-                             pos: Optional[int] = None) -> None:
-        if pos is None:
-            pos = len(self.coded_message)
+    def emplace_atomic_value(self, new_data: bytes, param_name: str) -> None:
+        pos = self.cursor_byte_position
 
         # Make blob longer if necessary
         min_length = pos + len(new_data)
         if len(self.coded_message) < min_length:
             self.coded_message.extend([0] * (min_length - len(self.coded_message)))
 
-        for byte_idx_val, byte_idx_rpc in enumerate(range(pos, pos + len(new_data))):
-            # insert byte value
-            if self.coded_message[byte_idx_rpc] & new_data[byte_idx_val] != 0:
-                warnings.warn(
-                    f"Object '{param_name}' overlaps with another parameter (bytes are already set)",
-                    OdxWarning,
-                    stacklevel=1,
+        for i in range(len(new_data)):
+            # insert new byte. this is pretty hacky: it will fail if
+            # the value to be inserted is bitwise "disjoint" from the
+            # value which is already in the PDU...
+            if self.coded_message[pos + i] & new_data[i] != 0:
+                odxraise(
```

Comment:
Don't change from warning to raise, or you will be break the PDX files I have.

Reply:
can you use the non-strict mode? if that is activated, it only logs a warning under the "odxtools" topic...

Reply:
(I think overlapping usually parameters are a bug in the PDX file (maybe except for NRC-CONST), and should thus be clearly flagged because if this is encountered the results are undefined, i.e., they might be surprising.)

Reply:
no, but a while ago, somebody [added a piece of code](https://github.com/mercedes-benz/odxtools/commit/b844c7f396cffd079e9b4fd6698eaeabe5275f78) that spits out a warning if they do. I think that it is sensible to turn this up a notch when we're paranoid about undefined behavior (that's how I interpret the strict mode). But I see your point: [reverted](https://github.com/mercedes-benz/odxtools/pull/291/commits/35fd6ce8bb57be0f9af8bbcaef5d8c8381c924dd)...